libia2/include/ia2.h

```diff
@@ -36,6 +36,12 @@
 
 #endif
 
+static uint32_t ia2_get_pkey() {
+  uint32_t pkru;
+  __asm__("rdpkru" : "=a"(pkru) : "a"(0), "d"(0), "c"(0));
```

Comment:
I believe this has to be `__asm__` `volatile` because the pkru value comes out of "nowhere" as far as the optimizer is concerned--`rdpkru` is impure in that it doesn't compute an output from the listed inputs to the inline asm statement.

Reply:
To support insecure mode, this should be `IA2_RDPKRU` rather than `"rdplru"` and initialize `pkru` to zero.

Reply:
We use an output, so we don't need volatile in this particular case. The compiler won't be able to discard this assembly statement because it needs the `pkru` output variable. The only time we _need_ volatile is if the assembly has side-effects not declared as outputs. So a wrpkru probably needs volatile, but reading doesn't.

Reply:
The concern isn't the compiler removing it entirely, but reordering it with respect to other operations because it considers it to be "pure". For example, it could hoist it above somewhere that we write the PKRU, even if the write is considered side-effecting, because the compiler doesn't know that the write and read are linked through the PKRU.